### PR TITLE
Block<T>.ProtocolVersion & BlockHeader.ProtocolVersion

### PR DESCRIPTION
**Pull request by dahlia**
This implements #1142.

### CODE REVIEW

**Comment by dahlia** [comment]
@planetarium/libplanet PTAL.

File: CHANGES.md [hunk 1]
```diff
@@ -38,11 +50,10 @@ To be released.
  -  Added `BlockExceedingTransactionsException` class.  [[#1104], [#1110]]
  -  Added `BlockChain<T>.GetStagedTransactionIds()` method.  [[#1089]]
  -  (Libplanet.RocksDBStore) Added `maxTotalWalSize`, `keepLogFileNum` and
-    `maxLogFileSize` parameters into `RocksDBStore` constructor.
+    `maxLogFileSize` parameters into `RocksDBStore()` constructor.
     [[#1065], [#1102], [#1132]]
  -  Added `Swarm<T>.BlockDemand` property.  [[#1068], [#1102]]
  -  Added `BlockDemand` struct.  [[#1068], [#1102]]
- -  Added `Block<T>.Header` property.  [[#1070], [#1102]]
```
Comment (limebell) [inline CHANGES.md:45]: Was this change intended?

File: Libplanet.Tests/Blocks/BlockTest.cs [hunk 2]
```diff
@@ -338,19 +338,22 @@ public void EvaluateActionsPerTx()
 
             Transaction<DumbAction>[] blockIdx2Txs =
             {
+                // Note that these timestamps in themselves does not have any meanings but are
+                // only arbitrary.  These purpose to make their evaluation order in a block
```
Comment (riemannulus) [inline Libplanet.Tests/Blocks/BlockTest.cs:342]: Is there any reason insert two space?

File: Libplanet/Blocks/Block.cs [hunk 3]
```diff
@@ -21,6 +21,11 @@ namespace Libplanet.Blocks
     public class Block<T>
         where T : IAction, new()
     {
+        /// <summary>
+        /// The most latest protocol version.
+        /// </summary>
+        public const int CurrentProtocolVersion = 1;
```
Comment (limebell) [inline Libplanet/Blocks/Block.cs:27]: Latest protocol version is okay with constant?

File: Libplanet/Blocks/Block.cs [hunk 4]
```diff
@@ -126,9 +135,11 @@ public Block(Bencodex.Types.Dictionary dict)
         {
         }
 
+        // FIXME: Should this be
```
Comment (limebell) [inline Libplanet/Blocks/Block.cs:138]: Something is missing I guess...

File: Libplanet.Tests/Blocks/BlockHeaderTest.cs [hunk 5]
```diff
@@ -12,13 +14,124 @@ public class BlockHeaderTest : IClassFixture<BlockFixture>
...
+                    BlockHeader.TimestampFormat,
+                    CultureInfo.InvariantCulture
+                ),
+                preEvaluationHash: TestUtils.GetRandomBytes(32).ToImmutableArray(),
+                stateRootHash: ImmutableArray<byte>.Empty
+            );
+
+            Assert.Throws<InvalidBlockProtocolVersionException>(() =>
+                header.Validate(DateTimeOffset.UtcNow)
+            );
```
Comment (riemannulus) [inline Libplanet.Tests/Blocks/BlockHeaderTest.cs:124]: Could we have a more test case where valid block protocol version? Like https://github.com/planetarium/libplanet/pull/1147/files#diff-b70ec4dc8f1efa4e59f3584b07849b9155f1f3d448a2e026a5d80e61c3de4bf5R152

File: Libplanet/Blocks/BlockHeader.cs [hunk 6]
```diff
@@ -196,6 +210,11 @@ public Bencodex.Types.Dictionary ToBencodex()
                 .Add(NonceKey, Nonce.ToArray())
                 .Add(HashKey, Hash.ToArray());
 
+            if (ProtocolVersion != 0)
+            {
+                dict = dict.Add(ProtocolVersionKey, ProtocolVersion);
```
Comment (limebell) [inline Libplanet/Blocks/BlockHeader.cs:215]: How about remove if statement and force to record protocol version afterward?

File: Libplanet/Blocks/Block.cs (hunk 4 above)
Comment (riemannulus) [inline Libplanet/Blocks/Block.cs:138]: It seems does not write the word to follow.

File: CHANGES.md (hunk 1 above)
Comment (dahlia) [reply CHANGES.md:45]: I just moved this line upper.

File: Libplanet/Blocks/Block.cs (hunk 3 above)
Comment (dahlia) [reply Libplanet/Blocks/Block.cs:27]: Yeah, this shouldn't be changed by applications, since it's for Libplanet-level protocol versioning.

File: Libplanet/Blocks/BlockHeader.cs (hunk 6 above)
Comment (dahlia) [reply Libplanet/Blocks/BlockHeader.cs:215]: It has to accept the field being omitted for backward compatibility (which is the point of the [spec][1]).

[1]: https://github.com/planetarium/libplanet/issues/1142

File: Libplanet/Blocks/Block.cs (hunk 3 above)
Comment (limebell) [reply Libplanet/Blocks/Block.cs:27]: Wording "current" is confusing to me. What this constant means?

File: Libplanet.Tests/Blocks/BlockTest.cs (hunk 2 above)
Comment (dahlia) [reply Libplanet.Tests/Blocks/BlockTest.cs:342]: AFAIK it's an [old convention][1], and I believe it improves readability on plain text (two spaces eventually will be reduced to one space when it's rendered anyway).

[1]: https://en.wikipedia.org/wiki/Sentence_spacing

File: Libplanet/Blocks/Block.cs (hunk 3 above)
Comment (dahlia) [reply Libplanet/Blocks/Block.cs:27]: It's the latest protocol version that this revision of Libplanet is aware of.

File: Libplanet.Tests/Blocks/BlockHeaderTest.cs (hunk 5 above)
Comment (dahlia) [reply Libplanet.Tests/Blocks/BlockHeaderTest.cs:124]: The link seems lost.

File: Libplanet/Blocks/Block.cs (hunk 3 above)
Comment (limebell) [reply Libplanet/Blocks/Block.cs:27]: Understood. Thx :)

File: Libplanet/Blocks/BlockHeader.cs (hunk 6 above)
Comment (limebell) [reply Libplanet/Blocks/BlockHeader.cs:215]: 👌

File: Libplanet/Blocks/Block.cs (hunk 4 above)
Comment (dahlia) [reply Libplanet/Blocks/Block.cs]: Completed the sentence.

Comment (dahlia) [reply Libplanet/Blocks/Block.cs]: Completed the sentence.

**Comment by dahlia** [comment]
@planetarium/libplanet Rebased.  PTAL.

File: Libplanet.Tests/Blocks/BlockHeaderTest.cs (hunk 5 above)
Comment (riemannulus) [reply Libplanet.Tests/Blocks/BlockHeaderTest.cs:124]: Like this https://github.com/planetarium/libplanet/blob/main/Libplanet.Tests/Blocks/BlockHeaderTest.cs#L38-L39

**Review by junorouse — DISMISSED** [review]
I understand that libplanet separately control the flow of actions through protocol version management (this patch) instead of creating a new action file (i.e., Buy {2,3,4}.cs), am I right? -- for proper patching of https://github.com/planetarium/lib9c/pull/248

If so, this PR looks OK for me. However, from an attacker's point of view, if an attacker has enough mining power, the block's previous version can continue to be mined even if the new protocol version introduces.

There are some mitigations for it:

1. Set the maximum number of blocks(=limit) allowed with the previous protocol version.
2. Like lib9c, use an authorized miner to introduce a new version.
3. Change to a consensus algorithm to that selecting the next block proposer randomly.
4. ...

File: Libplanet.Tests/Blocks/BlockHeaderTest.cs (hunk 5 above)
Comment (dahlia) [reply Libplanet.Tests/Blocks/BlockHeaderTest.cs:131]: Added:

```diff
--- a/Libplanet.Tests/Blocks/BlockHeaderTest.cs
+++ b/Libplanet.Tests/Blocks/BlockHeaderTest.cs
@@ -98,6 +98,13 @@ namespace Libplanet.Tests.Blocks
             Assert.Equal(expected, new BlockHeader(expected).ToBencodex());
         }
 
+        [Fact]
+        public void ValidateValidHeader()
+        {
+            _fx.Genesis.Header.Validate(DateTimeOffset.UtcNow);
+            _fx.Next.Header.Validate(DateTimeOffset.UtcNow);
+        }
+
         [Fact]
         public void ValidateProtocolVersion()
         {
```

**Comment by dahlia** [comment]
@junorouse Thanks for your review and suggestions!

> However, from an attacker's point of view, if an attacker has enough mining power, the block's previous version can continue to be mined even if the new protocol version introduces.

Even if we don't merge this PR, that attackers can outpace honest nodes is the critical problem in itself and we need to address it as soon as possible.  Therefore, IMHO we'd better deal with it in the separate issue/patches.

File: CHANGES.md [hunk 7]
```diff
@@ -99,7 +119,6 @@ To be released.
  -  `Swarm<T>` became not to fill states from trusted peers, because now states
     can be validated rather than trusted due to MPT.  [[#1117]]
  -  `HashDigest<SHA256>` became serializable.  [[#795], [#1125]]
-
```
Comment (moreal) [inline CHANGES.md:102]: Is this change intended? Of course, it seems not to affect to the rendered result because it is markdown.

**Review by riemannulus — APPROVED** [review]

File: CHANGES.md (hunk 7 above)
Comment (dahlia) [reply CHANGES.md:102]: No, it's removed by mistake.  I'm going to revert it.

Comment (dahlia) [reply CHANGES.md]: Removed.

**Review by moreal — APPROVED** [review]

**Review by longfin — APPROVED** [review]
